src/performance.rs

```diff
@@ -0,0 +1,200 @@
+use std::sync::{Arc, Mutex};
+use lazy_static::lazy_static;
+use std::time::Instant;
+use std::borrow::BorrowMut;
+
+pub struct Marker {
+    start: Instant,
+    id: usize,
+}
+
+impl Drop for Marker {
+    fn drop(&mut self) {
+        PERF_COUNTER.lock().unwrap().end(self)
+    }
+}
+
+pub struct Counter {
+    name: &'static str,
+    file: &'static str,
+    line: u32,
+
+    total_calls: usize,
+    calls_this_frame: usize,
+    total_time: f64,
+    time_this_frame: f64,
+}
+
+impl Counter {
+    pub fn new(name: &'static str,
+               file: &'static str,
+               line: u32) -> Self {
+        Self {
+            name,
+            file,
+            line,
+
+            total_calls: 0,
+            calls_this_frame: 0,
+            total_time: 0.0,
+            time_this_frame: 0.0,
+        }
+    }
+
+    fn add(&mut self, start: Instant) {
+        self.total_calls += 1;
+        self.calls_this_frame += 1;
+        let time = Instant::now().duration_since(start).as_secs_f64();
+        self.total_time += time;
+        self.time_this_frame += time;
+    }
+}
+
+#[derive(Copy, Clone, Debug)]
+pub enum CaptureWindow {
+    Nothing,
+    Silent,
+    CaptureFor(usize),
+    LogEvery(usize),
+    Everything,
+}
+
+impl CaptureWindow {
+    fn should_log(&self, frame: usize) -> bool {
+        match self {
+            CaptureWindow::Everything | CaptureWindow::CaptureFor(_) => true,
+            CaptureWindow::LogEvery(x) => (frame % x) == 0,
+            _ => false,
+        }
+    }
+
+    fn should_capture(&self) -> bool {
+        !matches!(self, CaptureWindow::Nothing)
+    }
+
+    fn step(self) -> Self {
+        match self {
+            CaptureWindow::CaptureFor(0) => CaptureWindow::Nothing,
+            CaptureWindow::CaptureFor(x) => CaptureWindow::CaptureFor(x - 1),
+            x => x,
+        }
+    }
+}
+
+pub fn capture_for(window: CaptureWindow) {
+    PERF_COUNTER.lock().unwrap().borrow_mut().window = window;
+}
+
+pub fn frame() {
+    let mut counter = PERF_COUNTER.lock().unwrap();
+    let counter = counter.borrow_mut();
+    if counter.window.should_capture() {
+        counter.frame();
+    }
+    if counter.window.should_log(counter.num_frames) {
+        counter.log();
+    }
+}
+
+#[macro_export]
+macro_rules! counter {
+    ( $name:expr ) => {
+        {
+            let info = lingon::performance::Counter::new(
+                $name,
+                std::file!(),
+                std::line!(),
+            );
+            let counter = lingon_macro::perf_counter!();
+            lingon::performance::PERF_COUNTER.lock().unwrap().start(counter, info)
+        }
+    };
+}
+
+lazy_static! {
+    pub static ref PERF_COUNTER: Arc<Mutex<Collector>> = Arc::new(Mutex::new(Collector::new()));
+}
```

Comment:
I think this should be at the top of this file. Since it technically is a global variable.